PR: https://github.com/odoo/odoo/pull/

From: 772bebede53ce75f4cc8682c8502920610d1d66f
From: jem-odoo
Date: 2018-08-13 12:33:48

Structural Changes: 3
Total Changes: 71

[MOV] project,hr_timesheet: move project_time_mode_id field

'project_time_mode_id' is the field used to determined the UoM to
use when logging timesheets. It is not used in project alone, but
in timesheet. Moving it make sense, and does not cause any problems.
For now, no UI allow the end user to change it: it is set with the
Hours UoM (hardcoded).
This is a technical move to prepare the feature that will allow
people timesheeting in Days or Hours.

Task #39079

================================= pseudo patch: =================================

--- a/addons/hr_timesheet/models/__init__.py
+++ b/addons/hr_timesheet/models/__init__.py
@@ -4,6 +4,7 @@
 from . import analytic_account
 from . import hr
 from . import hr_timesheet
+from . import res_company
 from . import res_config_settings
 from . import project
 from . import res_users

--- a/addons/hr_timesheet/models/project.py
+++ b/addons/hr_timesheet/models/project.py
@@ -1,6 +1,8 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
+from lxml import etree
+
 from odoo import models, fields, api, _
 from odoo.exceptions import UserError, ValidationError
 
@@ -151,3 +153,43 @@ class Task(models.Model):
                 'company_id': account_id.company_id.id,
             })
         return result
+
+    @api.model
+    def fields_view_get(self, view_id=None, view_type='form', toolbar=False, submenu=False):
+        # read uom as admin to avoid access rights issues, e.g. for portal/share users,
+        # this should be safe (no context passed to avoid side-effects)
+        obj_tm = self.env.user.company_id.project_time_mode_id
+        tm = obj_tm and obj_tm.name or 'Hours'
+
+        res = super(Task, self).fields_view_get(view_id=view_id, view_type=view_type, toolbar=toolbar, submenu=submenu)
+
+        # read uom as admin to avoid access rights issues, e.g. for portal/share users,
+        # this should be safe (no context passed to avoid side-effects)
+        obj_tm = self.env.user.company_id.project_time_mode_id
+        # using get_object to get translation value
+        uom_hour = self.env.ref('uom.product_uom_hour', False)
+        if not obj_tm or not uom_hour or obj_tm.id == uom_hour.id:
+            return res
+
+        eview = etree.fromstring(res['arch'])
+
+        # if the project_time_mode_id is not in hours (so in days), display it as a float field
+        def _check_rec(eview):
+            if eview.attrib.get('widget', '') == 'float_time':
+                eview.set('widget', 'float')
+            for child in eview:
+                _check_rec(child)
+            return True
+
+        _check_rec(eview)
+
+        res['arch'] = etree.tostring(eview, encoding='unicode')
+
+        # replace reference of 'Hours' to 'Day(s)'
+        for f in res['fields']:
+            # TODO this NOT work in different language than english
+            # the field 'Initially Planned Hours' should be replaced by 'Initially Planned Days'
+            # but string 'Initially Planned Days' is not available in translation
+            if 'Hours' in res['fields'][f]['string']:
+                res['fields'][f]['string'] = res['fields'][f]['string'].replace('Hours', obj_tm.name)
+        return res

--- a/None
+++ b/addons/hr_timesheet/models/res_company.py
@@ -0,0 +1,23 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models
+
+
+class ResCompany(models.Model):
+    _inherit = 'res.company'
+
+    @api.model
+    def _default_project_time_mode_id(self):
+        uom = self.env.ref('uom.product_uom_hour', raise_if_not_found=False)
+        if not uom:
+            uom = self.env['uom.uom'].search([('measure_type', '=', 'time'), ('uom_type', '=', 'reference')], limit=1)
+        if not uom:
+            uom = self.env['uom.uom'].search([('measure_type', '=', 'time')], limit=1)
+        return uom
+
+    project_time_mode_id = fields.Many2one('uom.uom', string='Project Time Unit',
+        default=_default_project_time_mode_id, domain=[('measure_type', '=', 'time')],
+        help="This will set the unit of measure used in projects and tasks.\n"
+             "If you use the timesheet linked to projects, don't "
+             "forget to setup the right unit of measure in your employees.")

--- a/addons/hr_timesheet/models/res_config_settings.py
+++ b/addons/hr_timesheet/models/res_config_settings.py
@@ -9,3 +9,8 @@ class ResConfigSettings(models.TransientModel):
 
     module_project_timesheet_synchro = fields.Boolean("Awesome Timesheet")
     module_project_timesheet_holidays = fields.Boolean("Leaves")
+    project_time_mode_id = fields.Many2one(
+        'uom.uom', related='company_id.project_time_mode_id', string='Project Time Unit',
+        help="This will set the unit of measure used in projects and tasks.\n"
+             "If you use the timesheet linked to projects, don't "
+             "forget to setup the right unit of measure in your employees.")
